Support for placing the drawer above the content.

mnexo opened the pull request:
Hey Simon/Jake

First of all, thanks for sharing your app menu drawer implementation.

I've extended the library a little, adding support for placing the drawer above the content, useful for playback controls or other status information.

You can see a use case from one of our apps here where the ongoing cooking timers in a recipe app are displayed in a drawer above the recipe: https://dl.dropbox.com/u/1288556/GoCookTimers.png.

Notes on the implementation:
- All fields now representing width or height (depending on the drawer's gravity) have
  been renamed from 'width' to 'size', e.g. mMenuWidth => mMenuSize.
- Added `MenuDrawer.MENU_POSITION_TOP` constant for use with the `gravity` argument in
  the constructor of `MenuDrawerManager`. Also refactored the creation of the drawer in
  the constructor into it's own helper method `createMenuDrawerForGravity`.
- Introduced HorizontalMenuDrawer specializing onMeasure and touch-related methods for
  horizontally oriented drawers (could use a better name).
- TopDrawer is simply a clone of LeftDrawer, swapping the x and y axes as appropriate.
- Access modifiers on a few fields and methods in `MenuDrawer` had to be changed from
  `private` to `protected` in order to specialize behavior in the new sub-classes.
- Due to lint warnings and errors with with MenuDrawer.postOnAnimate override (the 
  @TargetApi annotation doesn't seem to help with override methods..), I switched to
  using ViewCompat.onPostAnimate instead of the hand-rolled one. This, obviously, has 
  the side effect of making the library depend on the support library. If you would
  rather avoid that it should be fairly straightforward to roll back this change.
- I appologize for the many non-significant diffs (line wraps, formatting). I wasn't 
  able to set up my Eclipse formatter to fully match the rules you have. If you send
  me an Eclipse formatter I'll happily fall in line.
- Added a sample activity demonstrating the use of TopDrawer.

Let me know what you think. Thanks!
Michael Nexo

Commit:
Use ViewCompat's implementation of postOnAnimation

Though the existing implementation was fine, Android Lint was
complaining, even when @TargetApi(16) had been added. I believe this
was because the postOnAnimation method was an override.
Since a backport of postOnAnimation is now implemented in ViewCompat,
it is probably better to just use that.

Commit:
Renamed everything 'width' to 'size' in preparation for top drawer.

Commit:
Added support for placing the menu drawer at the top:

 - HorizontalMenuDrawer is a specialization of MenuDrawer specializing
   measure and touch events for revealing the menu at the top or bottom.
 - TopDrawer implements layout and manipulation of a menu placed at the
   top.
 - Added TopMenuSample activity showing it in action.

JakeWharton commented on library/src/net/simonvt/widget/MenuDrawer.java:
```diff
@@ -576,7 +588,7 @@ public void setOnDrawerStateChangeListener(OnDrawerStateChangeListener listener)
     /**
      * Defines whether the drop shadow is enabled.
-     *
+     * 
```
You need to set your editor to truncate trailing whitespace. There's a million of these...

JakeWharton commented on library/src/net/simonvt/widget/HorizontalMenuDrawer.java:
```diff
@@ -0,0 +1,201 @@
+
```
Extra line.

JakeWharton commented on library/src/net/simonvt/widget/LeftDrawer.java:
```diff
@@ -1,3 +1,4 @@
+
```
Extra line.

JakeWharton commented on library/src/net/simonvt/widget/MenuDrawerManager.java:
```diff
@@ -132,7 +150,8 @@ private void attachToDecor() {
         decorView.removeAllViews();
-        decorView.addView(mMenuDrawer, LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT);
+        decorView.addView(mMenuDrawer, LayoutParams.MATCH_PARENT,
+                LayoutParams.MATCH_PARENT);
```
There's a lot of unnecessary line wraps introduced. Make sure your IDE is configured to use the proper line width allowance.

JakeWharton commented on samples/src/net/simonvt/menudrawer/samples/TopMenuSample.java:
```diff
@@ -0,0 +1,99 @@
+ * Sample class illustrating how to add a menu drawer above the content area.
+ */
+public class TopMenuSample extends Activity implements OnClickListener
+{
```
Brace should be on above line.

JakeWharton replied:
You have a bunch of these... make sure your editor is configured properly.

mnexo replied:
Sorry, this project was simply set up with the wrong formatter...

JakeWharton commented on samples/src/net/simonvt/menudrawer/samples/TopMenuSample.java:
```diff
@@ -0,0 +1,99 @@
+	private TextView mContentTextView;
+
+	@Override
+	@TargetApi(16)
```
Not available with Maven compilation.

JakeWharton replied:
You have a bunch of these...

JakeWharton commented on samples/src/net/simonvt/menudrawer/samples/TopMenuSample.java:
```diff
@@ -0,0 +1,99 @@
+	{
+		MenuItem overflowItem = menu.add(0, MENU_OVERFLOW, 0, null);
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH)
+			overflowItem.setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
```
Use `MenuCompat` for this.

JakeWharton commented on samples/src/net/simonvt/menudrawer/samples/TopMenuSample.java:
```diff
@@ -0,0 +1,99 @@
+	{
+		switch (item.getItemId())
+		{
+		case android.R.id.home:
```
`case` statements need indented.

JakeWharton commented on library/src/net/simonvt/widget/TopDrawer.java:
```diff
@@ -0,0 +1,147 @@
+            animateOffsetTo(mVelocityTracker.getYVelocity() > 0 ? mMenuSize : 0, initialVelocity,
+                    true);
+
+            // Close the menu when content is clicked while the menu is visible.
```
Odd comment location. If it's for the `else if` put it inside as the first line

mnexo replied:
This is essentially copy/paste from the existing Left- and RightDrawer.java files which have the same comment (at the same location).

SimonVT replied:
@JakeWharton is right, the comment should be moved. If you do it here, I'll take care of other places in the library.

JakeWharton commented on library/src/net/simonvt/widget/TopDrawer.java:
```diff
@@ -0,0 +1,147 @@
+import android.util.AttributeSet;
+import android.view.MotionEvent;
+
+public class TopDrawer extends HorizontalMenuDrawer {
```
Odd class hierarchy. Why not extend `MenuDrawer` directly? If it does not have the functionality it needs we should refactor the common stuff up as high as possible.

mnexo replied:
This is in anticipation of a BottomDrawer. Top- and bottom drawers would both inherit from HorizontalMenuDrawer which contains the shared logic for these.

You could argue that onMeasure, onTouchEvent and onInterceptTouchEvent should be moved from MenuDrawer into a VerticalMenuDrawer as these methods contain code only needed for the left/right case.

SimonVT replied:
I kinda like the idea of having a Vertical/HorizontalMenuDrawer. Stuff like touch handling and measuring is quite different between the two types of implementations, but a lot of it is common between Top/Bottom and Left/Right. It makes sense to split it up.
I'm a bit confused why Top/BottomDrawer extends HorizontalMenuDrawer tho. Shouldn't it be Vertical? Since it opens up vertically.

JakeWharton commented on library/src/net/simonvt/widget/MenuDrawer.java:
```diff
@@ -998,10 +1021,11 @@ private void peekDrawerInvalidate() {
 
             if (!mPeekScroller.isFinished()) {
-                postOnAnimation(mPeekRunnable);
+                ViewCompat.postOnAnimation(this, mPeekRunnable);
```
I don't think this is available in the r7 support library (the only one available through Maven)

JakeWharton commented on library/src/net/simonvt/widget/MenuDrawer.java:
```diff
@@ -688,16 +702,18 @@ public void setHardwareLayerEnabled(boolean enabled) {
+            if (mOnDrawerStateChangeListener != null)
+                mOnDrawerStateChangeListener.onDrawerStateChange(oldState, state);
+            if (DEBUG)
+                logDrawerState(state);
```
Bad line wraps.

JakeWharton commented on library/src/net/simonvt/widget/MenuDrawer.java:
```diff
@@ -763,37 +779,39 @@ protected void dispatchDraw(Canvas canvas) {
+        if (mDropShadowEnabled)
+            drawDropShadow(canvas, offsetPixels);
+        if (mArrowBitmap != null)
+            drawArrow(canvas, offsetPixels);
```
Bad line wraps.

JakeWharton commented on library/src/net/simonvt/widget/MenuDrawer.java:
```diff
@@ -846,10 +865,12 @@ protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
+        if (!mMenuSizeSet)
+            mMenuSize = (int) (width * 0.8f);
+        if (mOffsetPixels == -1)
+            setOffsetPixels(mMenuSize);
```
Bad line wraps.

Commit:
A number of changes to make pull request more palatable:

 - Removed use of @TargetApi -- maven doesn't support it.
 - Removed use of support-lib from library project.
 - Minimized formatting changes (trailing whitespace, linebreaks, etc.)
 - Fixed maven checkstyle errors.

mnexo commented:
Thanks for the review, Jake.

I've committed changes for most -- the ones I haven't dealt with I've commented upon above.

Should I create a new pull request or somehow update this one? (this is my first pull-request on Github, so please pardon my ignorance...)

Thanks
Michael

SimonVT commented:
Hello, thanks for contributing.

You should rebase this against the dev branch. I pushed some changes a few days ago (deprecated MenuDrawerManager). It should be quite trivial to move the logic to the new MenuDrawer#attach(..) method.

I don't like the dependency on the support library, there's really no reason for it. I'm not too concerned about lint complaining about api level either, imo it would just increase the complexity to refactor code like this into static classes (which is the "correct" way to handle it).

Thanks

SimonVT commented:
Also, is it possible to squash the commits that fixes formatting in with the actual change? Just to clean up the noise before I review the changes, as it's easy to lose track otherwise.

mnexo commented:
Done. I'm going to close this pull request and open a new one based off of dev.

Rahulrsannidhi commented:
hi this menu drawer working smoothly.but i need to open the menu when apply long press in the any where of the screen now the menu is opening but when i leave the touch its closed (working when long press and swipe the screen)
